aiida_common_workflows/workflows/eos.py

```diff
@@ -0,0 +1,113 @@
+
+
+@calcfunction
+def scale_structure(structure: orm.StructureData, scale_factor: orm.Float) -> orm.StructureData:
+    """Scale the structure with the given scaling factor."""
+    ase = structure.get_ase().copy()
+    ase.set_cell(ase.get_cell() * float(scale_factor), scale_atoms=True)
+    return orm.StructureData(ase=ase)
+
+
+class EquationOfStateWorkChain(WorkChain):
+    """Workflow to compute the equation of state for a given crystal structure."""
+
+    @classmethod
+    def define(cls, spec):
+        # yapf: disable
+        super().define(spec)
+        spec.input('structure', valid_type=orm.StructureData, help='The structure at equilibrium volume.')
+        spec.input('scale_factors', valid_type=orm.List, required=False, validator=validate_scale_factors,
+            help='The list of scale factors at which the volume and total energy of the structure should be computed.')
+        spec.input('scale_count', valid_type=orm.Int, default=lambda: orm.Int(7), validator=validate_scale_count,
+            help='The number of points to compute for the equation of state.')
+        spec.input('scale_increment', valid_type=orm.Float, default=lambda: orm.Float(0.02),
+            validator=validate_scale_increment,
+            help='The relative difference between consecutive scaling factors.')
+        spec.input_namespace('sub_process', dynamic=True)
+        spec.input('sub_process_class', non_db=True, validator=validate_sub_process_class)
+        spec.inputs.validator = validate_inputs
+        spec.outline(
+            cls.run_eos,
+            cls.inspect_eos,
+        )
+        spec.output_namespace('structures', valid_type=orm.StructureData,
+            help='The relaxed structures at each scaling factor.')
+        spec.output_namespace('total_energies', valid_type=orm.Float,
+            help='The computed total energy of the relaxed structure at each scaling factor.')
+        spec.exit_code(400, 'ERROR_SUB_PROCESS_FAILED',
+            message='At least one of the `{cls}` sub processes did not finish successfully.')
+
+    def run_eos(self):
+        """Run the sub process at each scale factor to compute the structure volume and total energy."""
+        process_class = WorkflowFactory(self.inputs.sub_process_class)
+        process_class_name = process_class.__name__
+
+        if 'scale_factors' in self.inputs:
+            scale_factors = self.inputs.scale_factors
+        else:
+            count = self.inputs.scale_count.value
+            increment = self.inputs.scale_increment.value
+            scale_factors = [orm.Float(1 + i * increment - (count - 1) * increment / 2) for i in range(count)]
+
+        for scale_factor in scale_factors:
+            structure = scale_structure(self.inputs.structure, scale_factor)
+            self.report(f'submitting `{process_class_name}` for scale_factor `{scale_factor}`')
+            node = self.submit(process_class, structure=structure, **self.inputs.sub_process)
```

Comment:
Also noticing now that this approach does not work for cp2k. In fact the `Cp2kRelaxWorkChain` does not have an input port `structure`. it is `cp2k.structure`

Reply:
Hmm, good point. Is this the only implementation that doesn't expose the `structure` at the top level? Not sure how we can generalize this. I see that the reason they did this is probably because they only have the [`BaseWorkChain`](https://github.com/aiidateam/aiida-cp2k/blob/develop/aiida_cp2k/workchains/base.py) which also takes care of relaxing structures. There is no dedicated relax workchain.

Reply:
In the fleur case it is also not toplevel but in a namespace. Therefore, for Fleur currently this also does not work. 
I am thinking about it to change it from the Fleur side.